src/Common/src/Interop/Unix/System.Native/Interop.ReadDir.cs

```diff
@@ -27,76 +28,63 @@ internal enum NodeType : int
         }
 
         [StructLayout(LayoutKind.Sequential)]
-        private unsafe struct InternalDirectoryEntry
+        internal unsafe struct DirectoryEntry
         {
-            internal IntPtr     Name;
-            internal int        NameLength;
-            internal NodeType   InodeType;
-        }
+            internal byte* Name;
+            internal int NameLength;
+            internal NodeType InodeType;
 
-        internal struct DirectoryEntry
-        {
-            internal NodeType   InodeType;
-            internal string     InodeName;
+            internal ReadOnlySpan<char> GetName(Span<char> buffer)
+            {
+                Debug.Assert(buffer.Length >= Encoding.UTF8.GetMaxCharCount(255), "should have enough space for the max file name");
+                Debug.Assert(Name != null, "should not have a null name");
+
+                int byteLength = NameLength;
+                if (byteLength == -1)
+                {
+                    byteLength = 0;
```

Comment:
is this state of the art for strlen of a buffer? Ideally there would be something we could reuse, which would likely be designed to vectorize.

`SpanHelpers.IndexOf(ref first, 0, int32.MaxValue)`  ?

Reply:
@ahsonkhan ?

Reply:
Afaik. @jkotas, @KrzysztofCwalina thoughts?

Reply:
It depends on how large the buffer is and how quickly do you expect to hit the value. Vectorization doesn't really help for small buffer sizes.

You can try:
```C#
buffer.IndexOf<byte>(0)
```

Reply:
As an aside, you could convert this to a do while instead of a while loop to save 1-2 instructions.

```C#
if (byteLength == -1)
{
    byte* first = Name;
    do
    {
        byteLength++;
    } while (*first++ != 0);
}
```

Reply:
@ahsonkhan I suppose I can optimize this with what I know (https://linux.die.net/man/3/readdir) to be true in this case. We really just have a pointer to the name in the dirent struct that the OS allocated.